tests/std/tests/GH_003131_tuple_opeq_overload_and/test.cpp

```diff
@@ -0,0 +1,57 @@
+#include <cassert>
+#include <tuple>
+
+using namespace std;
+
+struct WeirdBoolBinOp {
+    bool b;
+
+    operator bool() const {
+        return b;
+    }
+    WeirdBoolBinOp operator!() const {
+        return WeirdBoolBinOp{!b};
+    }
+
+    // EVILLLLL
+    friend bool operator&&(const WeirdBoolBinOp& lhs, bool rhs) {
+        return !lhs.b && rhs;
+    }
+    friend bool operator||(const WeirdBoolBinOp& lhs, bool rhs) {
+        return !lhs.b || rhs;
+    }
+};
+struct WeirdBoolNot {
+    bool b;
+
+    operator bool() const {
+        return b;
+    }
+    // EVILLLLL
+    WeirdBoolNot operator!() const {

```

Comment:
Do we need two distinct types here? Why not just one type with `&&`, `||`, and `!` operators that explode?

Reply:
in theory they're two distinct problems, but yeah that's probably a good idea.